torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1265,20 +1265,56 @@ void ProcessGroupNCCL::terminateProcess(std::string errMsg) {
 
 void ProcessGroupNCCL::heartbeatMonitor() {
   uint64_t heartBeatCounter = 0ULL;
+  int heartBeatTimeout = (dumpOnTimeout_ && uid_ == 0)
+      ? timeoutCheckIntervalMilSec_ * 1000
+      : heartbeatTimeoutInSec_;
+  auto lastTimePollStore = std::chrono::steady_clock::now();
+  std::future<bool> asyncDebugDump;
   while (true) {
     // This won't have any lock since this lock is only used here.
     // Please be aware that mutex `monitorMutex_` should not be used
     // somewhere else to avoid the deadlock.
     std::unique_lock<std::mutex> lock(monitorMutex_);
     if (monitorWakeUpCV_.wait_for(
-            lock, std::chrono::seconds(heartbeatTimeoutInSec_), [&] {
+            lock, std::chrono::seconds(heartBeatTimeout), [&] {
               return terminateHeartbeatMonitorThread_.load();
             })) {
       // For the normal complete or user interception, monitorWakeUpCV_
       // will get notified, we early return and exit heartbeatMonitor.
       return;
     }
 
+    // Assuming that we always init a process group containing all ranks,
+    // we only use the default PG listen to the global signal to dump
+    // and abort. We poll store to see if some ranks have flagged a timeout when
+    // we haven't polled for `heartbeat_timeout` seconds and there haven't
+    // any work added or removed for `watchdog_timeout` seconds.
+    if (dumpOnTimeout_ && uid_ == 0) {
+      auto currentTime = std::chrono::steady_clock::now();
+      auto timeSinceLastWorkListUpdate =
+          std::chrono::duration_cast<std::chrono::milliseconds>(
+              (currentTime - lastWorkListUpdateTime_))
+              .count();
+      auto timeSinceLastPollStore =
+          std::chrono::duration_cast<std::chrono::milliseconds>(
+              (currentTime - lastTimePollStore))
+              .count();
+      if (timeSinceLastWorkListUpdate >= kWatchdogThreadSleepMillis &&
```

Comment:
this whole block is so hard to read.  i am trying to think of a way to make it briefer.

what if we made a helper for computeDeltaMS(start, end) = std::chrono::duration_cast<std::chrono::milliseconds>((end - start)).count()

and then we do 

if(computeDeltaMS(lastWorkListUpdateTime_ , currentTime) >= kWatchdogThreadSleepMillis &&
    computeDeltaMS(...) >=  ...) {